### PR TITLE
Added back CommonsJS output under /cjs

### PR DESCRIPTION
**Pull request by arthurfiorette**
https://github.com/vega/ts-json-schema-generator/pull/1922#issuecomment-2094493826 removed support for CommonsJS entirely without warnings and a major release. Since a lot of libraries depends on this package, adding a `cjs` bundle back should be considered.

You may argue that their libraries should migrate to ESM, however, one of the uses cases of this library is to generate json schemas inside a [Typescript Plugin](https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin) which is loaded by the TypeScript CJS source itself and will not be ported to ESM soon (if ever).

### CODE REVIEW

**Comment by arthurfiorette** [comment]
The main ts-json-schema-generator.ts file imported `package.json` to get its version, I was required to do that step as a post-process operation because I couldn't get the same ts source code transpiled into (cjs/esm) to work.

https://github.com/vega/ts-json-schema-generator/actions/runs/9105923115/job/25032310451#step:7:997

File: package.json [hunk 1]
```diff
@@ -44,13 +46,18 @@
...
+    "require": "./cjs/index.js"
+  },
   "dependencies": {
     "@types/json-schema": "^7.0.15",
     "commander": "^12.0.0",
     "glob": "^10.3.12",
     "json5": "^2.2.3",
     "normalize-path": "^3.0.0",
     "safe-stable-stringify": "^2.4.3",
+    "tslib": "^2.6.2",
```
Comment (domoritz) [inline package.json:60]: Why is this change needed?

File: ts-json-schema-generator.ts [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
-import { Config } from "./src/Config.js";
+import type { Config } from "./src/Config.js";
 import { BaseError } from "./src/Error/BaseError.js";
 import { formatError } from "./src/Utils/formatError.js";
-import pkg from "./package.json" with { type: "json" };
-import { dirname } from "path";
-import { mkdirSync, writeFileSync } from "fs";
+
+// This constant gets replaced by the build script
+const pkgVersion = "0.0.0"; /* __VERSION__ */
```
Comment (domoritz) [inline ts-json-schema-generator.ts]: I don't like this as it makes it harder to watch build. The cjs version used to work so can you try to make the import version work here as well?

**Review by domoritz — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (arthurfiorette) [reply package.json:60]: Insalling tslib and adding `importHelpers: true` in tsconfig drastically reduces build size:

https://www.typescriptlang.org/tsconfig/#importHelpers

File: ts-json-schema-generator.ts (hunk 2 above)
Comment (arthurfiorette) [reply ts-json-schema-generator.ts]: I tried a lot but ran out of choices.

- `import pkg from 'package.json'` (or * as pkg) works in cjs but not in ESM as it [requires](https://github.com/vega/ts-json-schema-generator/actions/runs/9105923115/job/25032310451#step:7:997) a `with {type: 'json'}` statement

- `require('package.json')` works but not in ESM mode.

- using `const _require = typeof require === 'undefined' ? createRequire(import.meta.url) : require` should work but breaks in CJS even if `createRequire(import.meta.url)` does not gets executed.

- using `fs.readFileSync` requires `__dirname` (which can be created using `import.meta.url` in esm but breaks just like the above try) and a second build step to manually copy package.json to `dist`/`cjs` folders.

- Trying to bypass `import.meta.url` using `eval('import.meta.url') does not work.

I'm not sure if there's any other way to fix it without replacing the dist code. Also, as it gets built on CI before publishing, all official releases should work just as fine.

File: package.json (hunk 1 above)
Comment (domoritz) [reply package.json:60]: I don't think we need to reduce build size in the same pull request.

Comment (arthurfiorette) [reply package.json:60]: `tslib` one of the most downloaded packages in npm. Everyone who installs one of its 120.000 dependents packages will have it already installed in their machine. 

As after this PR this package size will almost double (two js files, two js sourcemaps and two d.ts for each source file), enabling this tsconfig option could reduce the package download time and improve the CJS codebase in general.

Without `tslib` & `importHelpers`:

```js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ts = void 0;
__exportStar(require("./src/Error/BaseError.js"), exports);
```

With them:

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ts = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./src/Error/BaseError.js"), exports);
```

And it gets repeated with almost every commonsjs javascript file. 

This change does not affect any ESM user.

I can still remove it if you say so.

**Comment by arthurfiorette** [comment]
After all, the only thing I can say is:
 
 Welcome to the hell of maintaining ESM/CJS modularity in modern JavaScript.

File: package.json (hunk 1 above)
Comment (domoritz) [reply package.json:60]: Would it help for esm users? If so, let's move the import helpers to esm as well. Or is it not needed since esm has all the things we need? I'd still put it into the main tsconfig just because in the future we might use it.

File: package.json [hunk 3]
```diff
@@ -44,13 +46,18 @@
   "engines": {
     "node": ">=18.0.0"
   },
+  "exports": {
```
Comment (domoritz) [inline package.json:49]: Doesn't this prevent imports of specific sub-files? I think that's okay since we export most stuff from the main package but wanted to make sure that we are not preventing extensibility.

File: package.json (hunk 1 above)
Comment (arthurfiorette) [reply package.json:60]: Most of the things `tslib` helps is with import/export, but sure! It might help with some usecases.

**Comment by domoritz** [comment]
True. I've wasted too many days with bundling and packaging stuff.

File: package.json (hunk 3 above)
Comment (arthurfiorette) [reply package.json:49]: Is there an official way to test it?

Comment (arthurfiorette) [reply package.json:49]: I'm almost sure we would only disallow importing from other files if the exports field was defined as follows:

```
"exports": {
  ".": {
    "import": "./dist/index.js",
    "require": "./cjs/index.js"
  }
},
```

**Comment by arthurfiorette** [comment]
is there anything left?

**Comment by domoritz** [comment]
replaced by #1964
